server/session/handler_anvil.go

```diff
@@ -0,0 +1,224 @@
+					firstLevel = firstEnchant.Level()
+				}
+
+				resultLevel := max(firstLevel, e.Level())
+				if firstLevel == e.Level() {
+					resultLevel = firstLevel + 1
+				}
+
+				compatible := t.CompatibleWithItem(first.Item())
+				if _, ok := first.Item().(item.EnchantedBook); ok {
+					compatible = true
+				}
+
+				for _, e2 := range first.Enchantments() {
+					if t != e2.Type() && !t.CompatibleWithOther(e2.Type()) {
+						compatible = false
+						i++
+					}
+				}
+
+				if !compatible {
+					hasIncompatible = true
+					continue
+				}
+				hasCompatible = true
+
+				if resultLevel > t.MaxLevel() {
+					resultLevel = t.MaxLevel()
+				}
+				rarityCost := t.Rarity().ApplyCost
+				if enchant {
+					rarityCost = max(1, rarityCost/2)
+				}
+
+				result = result.WithEnchantments(item.NewEnchantment(t, resultLevel))
+				i += rarityCost * resultLevel
+				if first.Count() > 1 {
+					i = 40
+				}
+			}
+			if i == 0 && hasIncompatible && !hasCompatible {
+				return fmt.Errorf("no compatible enchantments but have incompatible ones")
+			}
+		}
+	}
+
+	newName := filterStrings[int(a.FilterStringIndex)]
+	existingName := item.DisplayName(first.Item(), s.c.Locale())
+	if customName := first.CustomName(); len(customName) > 0 {
+		existingName = customName
+	}
+	if existingName != newName {
+		k = 1
+		i += k
+		result = result.WithCustomName(newName)
+	}
+
+	cost := i + j
+	if cost <= 0 {
+		return fmt.Errorf("no action was taken")
+	}
+
+	if k == i && k > 0 && cost >= 40 {
+		cost = 39
+	}
+
+	c := s.c.GameMode().CreativeInventory()
+	if cost >= 40 && !c {
+		return fmt.Errorf("impossible cost")
+	}
+
+	if !result.Empty() {
+		k2 := result.RepairCost()
+		if !second.Empty() && k2 < second.RepairCost() {
+			k2 = second.RepairCost()
+		}
+		if k != i || k == 0 {
+			k2 = k2*2 + 1
+		}
+		result = result.WithRepairCost(k2)
+	}
+
+	level := s.c.ExperienceLevel()
+	if level < cost && !c {
+		return fmt.Errorf("not enough experience")
+	} else if !c {
+		s.c.SetExperienceLevel(level - cost)
+	}
+
+	if !c && rand.Float64() < 0.12 {
+		damaged := anvil.Break()
+		if _, ok := damaged.(block.Air); ok {
+			w.PlaySound(pos.Vec3Centre(), sound.AnvilBreak{})
+		} else {
+			w.PlaySound(pos.Vec3Centre(), sound.AnvilUse{})
+		}
+		defer w.SetBlock(pos, damaged, nil)
+	} else {
+		w.PlaySound(pos.Vec3Centre(), sound.AnvilUse{})
+	}
+
+	h.setItemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerAnvilInput,
+		Slot:        1,
+	}, item.Stack{}, s)
+	if repairCount > 0 {
+		h.setItemInSlot(protocol.StackRequestSlotInfo{
+			ContainerID: containerAnvilMaterial,
+			Slot:        2,
+		}, second.Grow(-repairCount), s)
+	} else {
+		h.setItemInSlot(protocol.StackRequestSlotInfo{
+			ContainerID: containerAnvilMaterial,
+			Slot:        2,
+		}, item.Stack{}, s)
+	}
+	h.setItemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerOutput,
+		Slot:        50,
+	}, result, s)
+	return nil
+}
```

Comment:
This is a massive function with a ton of logic in it. Consider splitting it up and particularly adding documentation inside of the code to explain what is going on. It's a really big wall of hard to comprehend logic for anyone else looking at it.

Reply:
Documented a lot of the code, just need to split it up into smaller functions. Will attempt this later today.

Reply:
Split up a few of the biggest blocks of code, please let me know if I should split up more.